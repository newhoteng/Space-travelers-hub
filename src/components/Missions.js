import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { getMissions } from '../redux/missions/missionsSlice';
import MissionItem from './MissionItem';
import styles from '../styles/Missions.module.css';

function Missions() {
  const { missions, isLoading, error } = useSelector((store) => store.missions);
  const dispatch = useDispatch();

  useEffect(() => {
    if (!missions.length) {
      dispatch(getMissions());
    }
  }, [dispatch, missions]);

  if (isLoading) {
    return (
      <ul>
        <h1>Loading...</h1>
      </ul>
    );
  }
  if (error) {
    return (
      <ul>
        <h1>Something went wrong</h1>
      </ul>
    );
  }

  return (
    <table className={styles.missionPage}>
      <thead>
        <tr>
          <th>Mission</th>
          <th>Description</th>
          <th>Status</th>
          <th className={styles.invisible}>Blank</th>
        </tr>
      </thead>
      <tbody>
        {missions.map((mission) => (
          <MissionItem
            key={mission.id}
            itemProps={mission}
          />
        ))}
      </tbody>
    </table>
  );
}

export default Missions;
