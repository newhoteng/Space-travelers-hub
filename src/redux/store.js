import { configureStore } from '@reduxjs/toolkit';
import missionsReducer from './missions/missionsSlice';
import dragonReducer from './Dragons/DragonsSlice';
import rocketReducer from './Rockets/RocketsSlice';

export const store = configureStore({
  reducer: {
    missions: missionsReducer,
    dragon: dragonReducer,
    rocket: rocketReducer,
  },
});

export default store;
