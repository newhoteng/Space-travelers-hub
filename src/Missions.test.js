import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { store } from './redux/store';

// Component imports
import Missions from './components/Missions';

import reducer, { joinMission, leaveMission } from './redux/missions/missionsSlice';

describe('Missions Slice', () => {
  test('should handle reserving a mission', () => {
    const state = {
      missions: [
        {
          id: '0', name: 'Thaicom', description: 'The best', reserved: false,
        },
        {
          id: '1', name: 'Telstar', description: 'Second best', reserved: false,
        },
        {
          id: '2', name: 'Iridium', description: 'Third best', reserved: false,
        },
      ],
      isLoading: false,
      error: undefined,
    };

    expect(reducer(state, joinMission('1'))).toEqual(
      {
        missions: [
          {
            id: '0', name: 'Thaicom', description: 'The best', reserved: false,
          },
          {
            id: '1', name: 'Telstar', description: 'Second best', reserved: true,
          },
          {
            id: '2', name: 'Iridium', description: 'Third best', reserved: false,
          },
        ],
        isLoading: false,
        error: undefined,
      },
    );
  });

  test('should handle cancelling a mission', () => {
    const state = {
      missions: [
        {
          id: '0', name: 'Thaicom', description: 'The best', reserved: true,
        },
        {
          id: '1', name: 'Telstar', description: 'Second best', reserved: false,
        },
        {
          id: '2', name: 'Iridium', description: 'Third best', reserved: false,
        },
      ],
      isLoading: false,
      error: undefined,
    };

    expect(reducer(state, leaveMission('0'))).toEqual(
      {
        missions: [
          {
            id: '0', name: 'Thaicom', description: 'The best', reserved: false,
          },
          {
            id: '1', name: 'Telstar', description: 'Second best', reserved: false,
          },
          {
            id: '2', name: 'Iridium', description: 'Third best', reserved: false,
          },
        ],
        isLoading: false,
        error: undefined,
      },
    );
  });
});

it('mission renders correctly', () => {
  const missions = renderer
    .create(
      <Provider store={store}>
        <Missions page="/missions">Missions Page</Missions>
      </Provider>,
    )
    .toJSON();
  expect(missions).toMatchSnapshot();
});

it('renders Missions component', () => {
  render(
    <Provider store={store}>
      <Missions />
    </Provider>,
  );
});
