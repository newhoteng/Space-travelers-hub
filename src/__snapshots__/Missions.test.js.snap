// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mission renders correctly 1`] = `
<table
  className="missionPage"
>
  <thead>
    <tr>
      <th>
        Mission
      </th>
      <th>
        Description
      </th>
      <th>
        Status
      </th>
      <th
        className="invisible"
      >
        Blank
      </th>
    </tr>
  </thead>
  <tbody />
</table>
`;
